Tutte polynomial

lucasmccabe opened the pull request:
This PR implements the Tutte polynomial, with space for additional graph polynomials to be included in the future.

Code has been run through black and flake8. All methods have been tested locally.

Thank you!

<!--
Please run black to format your code.
See https://networkx.org/documentation/latest/developer/contribute.html for details.
-->

lucasmccabe commented:
@dschult I would appreciate some help with this. I'm not understanding why the pypy installation keeps failing. Thank you!

rossbar commented:
The `Pypy` workflow is failing because it runs without the default dependencies. The imports of any dependencies should be moved within the calling function to prevent problems at import time. In addition to the `has_sympy` additions to the test configurations, you may also need to annotate the new tests with a `pytest.importorskip` to prevent failures during test collection when the default dependencies are not installed.

lucasmccabe commented:
@rossbar  Ah, I understand now - thank you for your help!

lucasmccabe commented:
Hello @rossbar - is this something you think we can consider for review, or do you think it is out of scope? Thank you!

dschult commented on networkx/algorithms/polynomials/tutte.py:
```diff
@@ -0,0 +1,91 @@
+"""Functions supporting the computation of the Tutte polynomial of a graph."""
```
You should write a concise one paragraph description of what a Tutte polynomial is. It can appear in the docs for this module, and if appropriate also in the doc_string of every appropriate function. Remember you reader is likely to not know what this is, and even if they do, they may have a different definition of it. 

The doc_strings should start with a short one-line description (that gets used to label any links in the docs to that function). It should be short -- like 80 chars or less) This must be followed by a blank line. Then put a description of what the function does in paragraph form. Define terms, give formulas, etc. Make it clear who should use this and perhaps more importantly who shouldn't use it.

You should also include references to articles or books or wikipedia with more info for the interested reader. Do you have publications where this subject is discussed?

lucasmccabe replied:
Good morning @dschult -  these comments make sense, and I've attempted to address them in my most recent commit. 

In particular, I provide an overview of the Tutte polynomial, some specializations and evaluations, three equivalent definitions, brief practical usage, and a few references for further reading.

Thank you for your help!

dschult commented on networkx/algorithms/polynomials/tutte.py:
```diff
@@ -0,0 +1,180 @@
+    """
+    cut_edges = []
+    for e in G.edges:
+        G_copy = G.copy()
+        G_copy.remove_edge(*e)
+        if len(list(nx.connected_components(G_copy))) > 1:
+            cut_edges.append(e)
```
This looks really really slow. Are you really going to copy the graph for every edge in the graph?
Also, you don't need to form the list of connected components if all you want is the number of them.
Try `nx.number_connected_components` instead.

Finally, why is your thinking behind making this a private function?

lucasmccabe replied:
Good point. I could just remove the edge and add it back, instead, e.g.:

```python
G_copy = G.copy()
cut_edges = []
for e in G.edges:
    G_copy.remove_edge(*e)
    if len(list(nx.connected_components(G_copy))) > 1:
        cut_edges.append(e)
    G_copy.add_edge(*e)
```

I can see why this should be a public function. Do you think it would make more sense to include in cuts module?

dschult commented on networkx/algorithms/polynomials/tutte.py:
```diff
@@ -0,0 +1,180 @@
+        return sympy.simplify(t)
+
+    cut_edges = _get_cut_edges(G)
+    loops = [e for e in G.edges if e[0] == e[1]]
```
Use `loops = set(nx.selfloop_edges(G))`  
The `set` choice is because in the next line we are checking if `i not in loop`.

lucasmccabe replied:
I'm happy to make a change in this direction, but I think `loops = list(nx.selfloop_edges(G, keys=True))` would be more appropriate, no? I may have multiple loops on the same node, and that is important for `sympy.Symbol("y") ** len(loops)`

dschult commented on networkx/algorithms/polynomials/tutte.py:
```diff
@@ -0,0 +1,180 @@
+    e = edges_not_cuts_loops[0]
+    # deletion-contraction
+    C = nx.contracted_edge(G, e, self_loops=True)
+    contraction_loops = [i for i in C.edges if i[0] == i[1] and i not in G.edges]
+    if contraction_loops:
+        C.remove_edge(*contraction_loops[0])
```
Shortcut this loop:           
```suggestion
    for e in C.edges if e[0] == e[1] and e not in G.edges:
        C.remove_edge(*e)
```
But actually --- why is this needed. Can you just use `self_loops=False` in the `contracted_edge` function?

lucasmccabe replied:
@dschult Good point, I will shortcut the loop as you describe.

This is needed because using `self_loops=False` in the `contracted_edge` function prohibits all loops arising from edge contraction. In this computation, we want to permit loops, but not arising from the particular edge we are contracting. The disagreement between these two approaches arises when you contract an edge with multiplicity greater than one.

For instance, suppose we have a three-vertex graph where one edge has multiplicity two:

```python
G = nx.MultiGraph()
G.add_edge(1, 2)
G.add_edge(1, 2)
G.add_edge(2, 3)
G.add_edge(1, 3)
```

Contracting one of the edges (1, 2) using `self_loops=False` in the `contracted_edge` function produces a graph with two vertices, two edges, and no loops. What we are looking for, however, is a graph with two vertices and three edges (one of which is a loop).

dschult commented on networkx/algorithms/polynomials/tutte.py:
```diff
@@ -0,0 +1,180 @@
+        C.remove_edge(*contraction_loops[0])
+    G.remove_edge(*e)
+
+    t = tutte_polynomial(G, False) + tutte_polynomial(C, False)
```
How big of a graph are you expecting this to work with? 
This recursion is likely to cause memory problems and run into recursion limits for graphs with many hundreds of nodes.  Can you do it without recursion?

lucasmccabe replied:
@dschult  @rossbar This is a good point and a known challenge. The recursive deletion-contraction method is the standard (and, until recently, fastest) general approach. Retrieving the coefficients is [#P-complete](https://dl.acm.org/doi/abs/10.1016/0166-218X(94)00097-W).

The fastest result I am aware of computes the coefficients in [vertex-exponential time](https://arxiv.org/abs/0711.2585). My intention was, if this feature were accepted as useful or interesting, to try implementing the vertex-exponential method as a companion function (similar to `gnp_random_graph` vs. `fast_gnp_random_graph`).

dschult replied:
Hmmm   I don't think recursive should impact the proof of speed of the algorithm. Any recursive algorithm can be rewritten as an iterative algorithm with a stack -- indeed that is what the programming language is doing in the background when you use recursion.  There is a stack of function over-heads that gets added to when another call is made and deleted from when each function call returns. 

The idea behind writing at as iterative (and one reason Python puts an adjustable recursion limit on it's kernels) is because the function calls are not a very good way to make and use a stack. Recursion is good for proving theorems, but it often is slow and a memory hog. 

I'm fine with including a recursive version if it is too hard to make one that is iterative. But it will limit the size of graphs, so I wondered what size you expect people to use with this algorithm?

dschult replied:
And "yes", it would be cool to write the other algorithm -- though if it is both fast and memory efficient, maybe it's best to make it the main function with the recursive version available for pedagogical value.

lucasmccabe replied:
@dschult Thank you for your careful review and explanation. I see now what you mean about writing at as iterative - I had misunderstood.

I have replaced the algorithm with a more practical iterative approach and relegated the recursive implementation to `tutte_polynomial_recursive` for pedagogical purposes. Is this preferable?

Thanks again!

lucasmccabe replied:
@dschult  Thank you for your careful review and explanation. I see now what you mean about rewriting as iterative - I had misunderstood.

> The idea behind writing at as iterative (and one reason Python puts an adjustable recursion limit on it's kernels) is because the function calls are not a very good way to make and use a stack. Recursion is good for proving theorems, but it often is slow and a memory hog.

This makes sense, thanks. In my most recent commit, `tutte_polynomial` is rewritten as an iterative algorithm with a stack, as you describe. The recursive implementation now appears in `tutte_polynomial_recursive` for pedagogical purposes.

> I'm fine with including a recursive version if it is too hard to make one that is iterative. But it will limit the size of graphs, so I wondered what size you expect people to use with this algorithm?

I don't expect this to be used for graphs with more than, say, a dozen vertices. More than that would take several hours, and even the most performant general implementation I'm aware of indicates a limit of 25 vertices ([Section D.2](https://arxiv.org/pdf/0711.2585.pdf)).

Luckily, as far as I'm aware, the graphs of interest for practical computations are small, e.g. for [biomolecular computing](https://link.springer.com/chapter/10.1007/978-3-319-08019-2_7).

lucasmccabe commented:
Hello @dschult @rossbar:

- In this most recent commit, `tutte_polynomial` employs an iterative version of the deletion-contraction algorithm, as recommended by @dschult.
- The original recursive implementation now appears in `tutte_polynomial_recursive` for pedagogical purposes.
- I have also added `tutte_polynomial_potts`, an alternative approach that obtains the Tutte polynomial via the partition sum of the q-state Potts model. This is not more performant, but may be of academic interest.

Thank you, as always!

lucasmccabe commented on networkx/algorithms/cuts.py:
```diff
@@ -390,3 +391,33 @@ def boundary_expansion(G, S):
     return len(nx.node_boundary(G, S)) / len(S)
+
+
+def cut_edges(G):
+    """Finds the cut-edges of a provided graph.
+
```
Thanks for pointing this out!

I am happy to switch to `nx.bridges`, but the algorithm requires a simple graph. My PR #5397 modifies it to accommodate multigraphs.

lucasmccabe commented on requirements/default.txt:
```diff
@@ -2,3 +2,5 @@ numpy>=1.19
 scipy>=1.8
 matplotlib>=3.4
 pandas>=1.3
+mpmath>=1.2
```
Good catch, thanks!

lucasmccabe commented on networkx/conftest.py:
```diff
@@ -350,6 +357,11 @@ def add_nx(doctest_namespace):
 needs_pydot = ["drawing/nx_pydot.py"]
 needs_ogr = ["readwrite/nx_shp.py"]
+needs_sympy = [
+    "algorithms/polynomials/__init__.py",
+    "algorithms/polynomials/tutte.py",
+    "algorithms/polynomials/tests/test_tutte.py",
```
Good catch, thanks!

lucasmccabe commented:
Hi @dschult - thank you for taking the time to review and provide these insights.

> Are sympy expressions the right data structure to represent polynomials?

Great question. In my view, yes.

It's true that using, say, a triangular matrix of coefficients may be faster if doing so does not otherwise change the algorithm (though only marginally, since the actual size of computable graphs is relatively small). For many of the fast computation techniques, however, the matrix approach is impractical. For instance, although I implement a classical algorithm here, I've mentioned elsewhere in our conversation that a [vertex-exponential](https://arxiv.org/abs/0711.2585) result has been proven possible. An efficient implementation of their idea warrants things like keeping terms symbolic until the end (e.g. the q from the q-state Potts model), "backing in" to the Tutte polynomial via other polynomials, etc. SymPy is great for this.

Also, SymPy provides a great deal of functionality for polynomial manipulation, and this is critical for realistic downstream tasks. One of the most interesting things about the Tutte polynomial is how many other graph polynomials and combinatorial properties are specializations; by using a SymPy expression, we can obtain specializations more [easily and quickly](https://docs.SymPy.org/latest/modules/numeric-computation.html#subs-evalf).

Of course, this is a design decision. But I think the utility and legibility of the SymPy approach is more in line with my understanding of the goals of NetworkX in general, e.g. prioritizing interpretability, utility for downstream analysis tasks, and ease of interfacing with other functions.


> You have made this a new directory within algorithms. Does that mean you expect to add more polynomial graph algorithms to networkx? 

Yes, this is my intent.

> It is usual to start with a single module and then when that gets too big, refactor it into a directory (subpackage). 

I see, thanks for letting me know. If I understand correctly, I think I've addressed this appropriately in my most recent commit.

> It's not clear why the potts function is included here, and why the function _partition_sum isn't just included as part of the potts function. (also the name _partition_sum is too generic -- but putting it inline into the potts function removes that difficulty. But more importantly, why are you including the potts algorithm? It looks like the loop for f in chain.from_iterable(combinations(edges, r) for r in range(len(edges) + 1)): could have really bad complexity as the number of edges gets large.

Agreed on all counts. This Potts approach is not better; I just thought it was interesting! I've removed it in the most recent commit.

> Thanks for rewriting the recursive version as an iterative version! It is actually very clean and doesn't complicate the code -- I would even say it is easier to read than the recursive version. Is there a reason to include the recursive version? If it is harder to read, then we should probably not include it for pedagogical value. :{

Good point - I think the iterative version may actually read better! I've removed the recursive version in the most recent commit.

> Combining all of these reductions may make this PR small enough to be easy to review. It could also provide a nice module where adding new polynomial algorithms would be easy for newcomers and maintainers alike.

Great! Thank you for your help. Graph polynomials are exciting and fun :)

dschult commented on networkx/algorithms/cuts.py:
```diff
@@ -390,3 +391,33 @@ def boundary_expansion(G, S):
+       "Introduction to Graph Theory," p. 23
+    """
+    G_copy = G.copy()
+    cut_edges = []
+    for e in G.edges:
+        G_copy.remove_edge(*e)
+        if nx.number_connected_components(G_copy) > nx.number_connected_components(G):
```
```suggestion
    ncc = nx.number_connected_components(G)
    cut_edges = []
    for e in G.edges:
        G_copy.remove_edge(*e)
        if nx.number_connected_components(G_copy) > ncc:
```

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,145 @@
+    There are several equivalent definitions; here are three:
+
+    Def 1 (rank-nullity expansion): For `G` an undirected graph, `n(G)` the
+    number of vertices of `G`, `E` the edge set of `G`, and `c(A)`` the number of
+    connected components of `A` [3]_:
```
In the formula, `A` is a subset of edges, but `c(A)` is the number of connected components of `A`. What is the graph used in `c(A)`?

lucasmccabe replied:
Good point. `c(A)` counts the connected components of the graph with the edge set `A` and the same vertex set as `G`. I will clarify this.

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,145 @@
+        T_G(x, y) = \sum_{A \in E} (x-1)^{c(A) - c(G)} (y-1)^{c(A) + |A| - n(G)}
+
+    Def 2 (spanning tree expansion): For `G` an undirected graph, `T` a spanning
+    tree of `G`, `i(T)` the internal activity of `T`, and `e(T)` the external
+    activity of `T` [4]_:
```
What are the internal activity and the external activity of a tree?

lucasmccabe replied:
I have updated the spanning tree expansion definition to include the following:

> Let `E` have an arbitrary strict linear order `L`. Let `B_e` be the unique minimal nonempty edge cut of `E \setminus T \cup {e}`. An edge `e` is internally active with respect to `T` and `L` if `e` is the least edge in `B_e` according to the linear order `L`. The internal activity of `T` (denoted `i(T)`) is the number of edges in `E \setminus T` that are internally active with respect to `T` and `L`. Let `P_e` be the unique path in `T \cup {e}` whose source and target vertex are the same. An edge `e` is externally active with respect to `T` and `L` if `e` is the least edge in `P_e` according to the linear order `L`. The external activity of `T` (denoted `e(T)`) is the number of edges in `E \setminus T` that are externally active with respect to `T` and `L`.

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,145 @@
+    Def 3 (deletion-contraction recurrence): For `G` an undirected graph, `G-e`
+    the graph obtained from `G` by deleting edge `e`, `G/e` the graph obtained
+    from `G` by contracting edge `e`, `k(G)` the number of cut-edges of `G`,
+    and `l(G)` the number of loops of `G`:
```
In NetworkX we call edges from a node to itself `selfloops` instead of `loops`.  I know that in some places `loops` is standard terminology, but we like the perhaps redundant reminder that these loops involve only one node.

lucasmccabe replied:
Good idea - I'll make the adjustment.

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,145 @@
+    .. math::
+        T_G(x, y) = \begin{cases}
+    	   x^{k(G)} y^{l(G)}, & \text{if all edges are cut-edges or loops} \\
+           T_{G-e}(x, y) + T_{G/e}(x, y), & \text{otherwise, for $e$ not a cut-edge or loop}
```
```suggestion
           T_{G-e}(x, y) + T_{G/e}(x, y), & \text{otherwise, for an arbitrary edge $e$ not a cut-edge or loop}
```

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,145 @@
+        G = stack.pop()
+        bridges = nx.cut_edges(G)
+        loops = list(nx.selfloop_edges(G, keys=True))
+        edges_not_cuts_loops = [
+            i for i in G.edges if i not in bridges and i not in loops
+        ]
```
If I read this correctly, you don't actually need all the `edges_not_cuts_loops` list... only the first element.
Can you loop over edges and break when you find one that works (and not loop over the others)?

Also, checking for a loop might be easier/thinner with `i not in loop` replaced by `i[0] != i[1]`
And once `nx.cut_edges` gets replaced by bridges, you should make it a set to speed the lookup `i not in bridges`.
Finally, `C.remove_edge(*(e[0], e[0]))` is more simply written `C.remove_edge(e[0], e[0])`.

lucasmccabe replied:
These are great ideas. I have made the updates in my most recent commit.

dschult reviewed (commented):
I've gone through this code and while I know changes will happen once #5397, I think these are the only concerns I have other than the cut_edges function which that other PR will handle.  Nice work...

lucasmccabe commented:
Good evening @dschult - thank you for your careful review and suggestions. I have incorporated these into my most recent commit.

Still remaining is to swap out `nx.cut_edges` for `nx.bridges`, which I will do as soon as #5397 is resolved.

Thank you for your help!

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,161 @@
+        G = stack.pop()
+        bridges = set(nx.cut_edges(G))
+
+        e = None
+        for i in G.edges:
+            if i not in bridges and i[0] != i[1]:
+                e = i
+                break
```
```suggestion
        for e in G.edges:
            if e not in bridges and e[0] != e[1]:
                break
```
Not a big deal, but slightly simpler.

lucasmccabe replied:
Hi @dschult - I'm not sure I agree with this one. This change would fail all tests.

I want to store the first edge that is not a bridge or self-loop. If I use `e` as the iterator variable, it will either be this or the last edge of `G.edges`. This will be problematic for line 151, where I check if all the remaining edges of `G` are bridges or self-loops.

dschult replied:
Yes, you are right!  You need to store `e=None` separately as an indicator of whether there were any or not.  Thanks!

lucasmccabe commented:
Good morning @dschult - per our earlier conversation, this commit swaps out `nx.cut_edges` for `nx.bridges`. Thank you!

dschult commented on networkx/algorithms/polynomials.py:
```diff
@@ -0,0 +1,161 @@
+
+@not_implemented_for("directed")
+def tutte_polynomial(G):
+    r"""Compute a graph's Tutte polynomial via an iterative version of the
+    deletion-contraction algorithm.
```
```suggestion
    r"""Return the Tutte polynomial of `G`
    
    This function computes the Tutte polynomial via an iterative version of
    the deletion-contraction algorithm.
```

lucasmccabe replied:
Good point; that hadn't occurred to me!

dschult reviewed (approved):
I approve this PR.  One nitpik suggstion to make the first line shorter (used to describe links in the documentation).  Up to you whether to accept or morph or leave as it is.

rossbar reviewed (approved):
Very nice @lucasmccabe thanks! I took the liberty of making a few final touchups (e.g. making sure the new polynomial module + fn shows up in the reference guide, minor rst formatting fixes for the docstring, etc.) but this LGTM. Thanks for all the hard work!

I'll let this sit for a while in case anyone wants to check the final few commits I pushed up. If it's not merged by tomorrow I'll push the green button!

lucasmccabe commented:
> Very nice @lucasmccabe thanks! I took the liberty of making a few final touchups (e.g. making sure the new polynomial module + fn shows up in the reference guide, minor rst formatting fixes for the docstring, etc.) but this LGTM. Thanks for all the hard work!

Thank you for these touch-ups - they look good to me. Thanks for your help @rossbar!
